File: src/frontend/src/services/dataQuality.js
// [AI GENERATED]
// Model: GitHub Copilot (GPT-4.1)
// Logic: Scaffold a frontend service for Data Quality Monitoring API endpoints.
// Why: Enables the React dashboard to report, retrieve, and display data quality events and alerts.
// Root Cause: No frontend integration existed for the new backend API.
// Context: This service provides methods for posting events, fetching all events, and fetching high-severity alerts. Future: add error handling, persistent storage, and advanced filtering. For advanced analytics, consider using a more advanced model (Claude Opus).

import axios from 'axios';

const API_BASE = '/api';

export const reportDataQualityEvent = async (event) => {
    try {
        const res = await axios.post(`${API_BASE}/data-quality/events`, event);
        return res.data;
    } catch (error) {
        console.error('Error reporting data quality event:', error);
        throw new Error('Failed to report data quality event. Please try again later.');
    }
};

export const getDataQualityEvents = async (dataset) => {
    if (dataset && typeof dataset !== 'string') {
        throw new Error('Invalid dataset parameter. It must be a string.');
    }

    try {
        const url = dataset ? `${API_BASE}/data-quality/events?dataset=${encodeURIComponent(dataset)}` : `${API_BASE}/data-quality/events`;
        const res = await axios.get(url);
        return res.data;
    } catch (error) {
        console.error('Error fetching data quality events:', error);
        throw new Error('Failed to fetch data quality events. Please try again later.');
    }
};

export const getHighSeverityAlerts = async () => {
    try {
        const res = await axios.get(`${API_BASE}/data-quality/alerts`);
        return res.data;
    } catch (error) {
        console.error('Error fetching high severity alerts:', error);
        throw new Error('Failed to fetch high severity alerts. Please try again later.');
    }
};

File: backend/core/routers/data_quality.py
# [AI GENERATED]
# Model: GitHub Copilot (GPT-4.1)
# Logic: Scaffold a FastAPI router for Data Quality Monitoring, providing endpoints for reporting, retrieving, and alerting on data drift and schema changes.
# Why: Roadmap item 21 requires proactive alerting for data drift and schema changes to ensure data reliability and compliance.
# Root Cause: No API existed for tracking or alerting on data quality issues.
# Context: This router provides endpoints for submitting quality metrics, retrieving drift reports, and triggering alerts. Future: integrate with Prometheus, add persistent storage, and support custom rules. For advanced anomaly detection, consider using a more advanced model (Claude Opus).

from datetime import datetime
from typing import List, Optional

from fastapi import APIRouter, status
from pydantic import BaseModel

router = APIRouter()

# In-memory store for demo purposes
DATA_QUALITY_EVENTS = []


class DataQualityEvent(BaseModel):
    id: int
    timestamp: datetime
    dataset: str
    event_type: str  # e.g., 'drift', 'schema_change', 'missing_values'
    details: Optional[str] = None
    severity: str  # e.g., 'low', 'medium', 'high'


@router.post("/data-quality/events", status_code=status.HTTP_201_CREATED)
def report_data_quality_event(event: DataQualityEvent):
    DATA_QUALITY_EVENTS.append(event)
    return {"message": "Event recorded", "event_id": event.id}


@router.get("/data-quality/events", response_model=List[DataQualityEvent])
def get_data_quality_events(dataset: Optional[str] = None):
    if dataset:
        return [e for e in DATA_QUALITY_EVENTS if e.dataset == dataset]
    return DATA_QUALITY_EVENTS


@router.get("/data-quality/alerts", response_model=List[DataQualityEvent])
def get_high_severity_alerts():
    return [e for e in DATA_QUALITY_EVENTS if e.severity == "high"]
